Add battery-notifier recipe

jasonmj opened the pull request:
### Brief summary of what the package does

This global minor mode sends notifications when battery capacity is low and suspends the computer when battery capacity is critically low.

### Direct link to the package repository

https://github.com/jasonmj/battery-notifier

### Your association with the package

Author and maintainer

### Relevant communications with the upstream package maintainer

This is a new package.

### Checklist

<!-- Please confirm with `x`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

<!-- After submitting, please fix any problems the CI reports. -->

purcell commented:
A couple of drive-by notes:
- You could use the `alert` package instead of providing your own way to customize how notifications are presented to the user. It's pretty great, and allows the options of window system popups etc.
- Instead of baking auto-suspend into the package, particularly in a way that is specific to a single OS platform, I'd suggest it would be more idiomatic to provide a `battery-notifier-low-level-hook`, and let the user add a function to it. At the very least there should be a way to disable this behaviour, e.g. and it should probably be off by default. You'd probably want to allow either of the threshold vars to be `nil`, indicating no action is wanted.
- [This check](https://github.com/jasonmj/battery-notifier/blob/ffe70fc1c951670248d553e01838187ac191a3ac/battery-notifier.el#L86) (and the other `bound-and-true-p` later) is nonsensical: of course the mode symbol will be bound! Just check its value. Also, you need to handle the case where it is nil, and suspend the watching, otherwise it's not really a toggle-able minor mode.
- [Here](https://github.com/jasonmj/battery-notifier/blob/ffe70fc1c951670248d553e01838187ac191a3ac/battery-notifier.el#L84) there is no such group defined: you should add a `defgroup` to declare and place your custom group in the customization hierarchy.

Hope that helps!

jasonmj commented:
Thanks for the quick and thoughtful feedback! 

> You could use the alert package instead of providing your own way to customize how notifications are presented to the user. It's pretty great, and allows the options of window system popups etc.

I had originally used alert, but thought it might be less intrusive to stick with message and let folks customize. At your suggestion though, I have incorporated alert. 

>  Instead of baking auto-suspend into the package, particularly in a way that is specific to a single OS platform, I'd suggest it would be more idiomatic to provide a battery-notifier-low-level-hook, and let the user add a function to it. At the very least there should be a way to disable this behaviour, e.g. and it should probably be off by default. 

I've replaced the suspend function with a customizable hook that is nil by default. 


> You'd probably want to allow either of the threshold vars to be nil, indicating no action is wanted.

I've added choices to the threshold customization variables that allow for them to be nil, indicating (and logically behaving) such that no action is taken when nil. 

> This check (and the other bound-and-true-p later) is nonsensical: of course the mode symbol will be bound! Just check its value. Also, you need to handle the case where it is nil, and suspend the watching, otherwise it's not really a toggle-able minor mode.

I've resolved  these issues by checking the mode value and canceling the timer when the mode is disabled. 

> Here there is no such group defined: you should add a defgroup to declare and place your custom group in the customization hierarchy.

I did have defgroup in there, but it wasn't above the mode definition. I've since rearranged the code and added more clear comments to improve this.

purcell commented:
Cool, looking good, sent you a few small fixes in https://github.com/jasonmj/battery-notifier/pull/1

jasonmj commented:
Thanks. I've approved and merged your changes. Let me know if there's anything else you need on my end. Very much appreciate you quick and succinct feedback. 😊

riscy commented:
Hi @jasonmj, I just have a few more quick lints but otherwise this is looking good --

- Add a LICENSE file that GitHub can detect (e.g. no markup) if possible
  See: https://github.com/licensee/licensee
- battery-notifier.el#L31: Use empty lines instead of `;;` on a line by itself within your Commentary
- battery-notifier.el#L129: You can use `not` or `null`
- battery-notifier.el#L134: You can use `not` or `null`

purcell commented:
I've merged for now to avoid too many passes, but yes, please follow up with the few little points raised above in your own time.

Welcome to MELPA! :tada:

jasonmj commented:
Thanks! I've addressed the issues raised and released a new version. Very pleased with this overall experience. I had expected to wait at least a week to get a response on this PR. Ya'll are awesome!

purcell commented:
Response times are a bit random, sorry, so it could easily have taken longer, and might do for any subsequent submissions. In this case I had a few minutes free and this package caught my eye, but generally we try to handle PRs in arrival order. :-)
